cobalt/browser/migrate_storage_record/migration_manager.cc

```diff
@@ -0,0 +1,314 @@
+    base::OnceClosure callback,
+    net::CookieAccessResult /* set_cookie_result */) {
+  if (cookies.empty()) {
+    GetCookieManager(weak_document_ptr)->FlushCookieStore(std::move(callback));
+    return;
+  }
+
+  auto cookie = std::move(cookies.back());
+  cookies.pop_back();
+  GURL source_url("https://" + cookie->Domain() + cookie->Path());
+  GetCookieManager(weak_document_ptr)
+      ->SetCanonicalCookie(
+          *cookie, source_url, net::CookieOptions::MakeAllInclusive(),
+          base::BindOnce(&MigrateCookiesHelper, weak_document_ptr,
+                         std::move(cookies), std::move(callback)));
+}
+
+void MigrateCookies(content::WeakDocumentPtr weak_document_ptr,
+                    std::vector<std::unique_ptr<net::CanonicalCookie>> cookies,
+                    base::OnceClosure callback) {
+  if (cookies.empty()) {
+    std::move(callback).Run();
+    return;
+  }
+  MigrateCookiesHelper(weak_document_ptr, std::move(cookies),
+                       std::move(callback), net::CookieAccessResult());
+}
+
+void MigrateLocalStorageHelper(
+    content::WeakDocumentPtr weak_document_ptr,
+    std::vector<std::pair<std::string, std::string>> entries,
+    base::OnceClosure callback,
+    base::Value /* result */) {
+  if (entries.empty()) {
+    std::move(callback).Run();
+    return;
+  }
+
+  auto* render_frame_host = weak_document_ptr.AsRenderFrameHostIfValid();
+  CHECK(render_frame_host);
+
+  std::string key = entries.back().first;
+  std::string value = entries.back().second;
+  entries.pop_back();
+  base::Value::List arguments = base::Value::List().Append(key).Append(value);
+  render_frame_host->ExecuteJavaScriptMethod(
+      base::ASCIIToUTF16(std::string("localStorage")),
+      base::ASCIIToUTF16(std::string("setItem")), std::move(arguments),
+      base::BindOnce(&MigrateLocalStorageHelper, weak_document_ptr,
+                     std::move(entries), std::move(callback)));
+}
+
+void MigrateLocalStorage(
+    content::WeakDocumentPtr weak_document_ptr,
+    std::vector<std::pair<std::string, std::string>> entries,
+    base::OnceClosure callback) {
+  MigrateLocalStorageHelper(weak_document_ptr, std::move(entries),
+                            std::move(callback), base::Value());
+}
+
+void Reload(content::WeakDocumentPtr weak_document_ptr,
+            base::OnceClosure callback) {
+  auto* render_frame_host = weak_document_ptr.AsRenderFrameHostIfValid();
+  CHECK(render_frame_host);
+  render_frame_host->Reload();
+  std::move(callback).Run();
+}
+
+}  // namespace
+
+// static
+std::atomic<bool> MigrationManager::migration_attempted_{false};
+
+// TODO(b/399165612): Add metrics.
+// TODO(b/399165796): Disable and delete migration code when possible.
+// TODO(b/399166308): Add unit and/or integration tests for migration.
+// static
+void MigrationManager::DoMigrationTasksOnce(
+    content::WebContents* web_contents) {
+  if (migration_attempted_.load()) {
```

Comment:
Ah I remember one comment: try to use `std::atomic_flag`, which has the neat operation `test_and_set`. It should be initialized like [this](https://en.cppreference.com/w/cpp/atomic/ATOMIC_FLAG_INIT):

` std::atomic_flag MigrationManager::migration_attempted_ = ATOMIC_FLAG_INIT;`

And then these lines l.276-279 would be:
```
const bool migration_attempted_before = migration_attempted_.test_and_set();
if (migration_attempted_before) {
  return;
}
```
which I think it's easier to read. (`test_and_set()` returns the value _before_ setting it).